orderbook/src/orderbook.rs

```diff
@@ -123,13 +124,54 @@ async fn set_order_balance(orders: &mut [Order], balance_fetcher: &dyn BalanceFe
     }
 }
 
-fn remove_orders_without_sufficient_balance(orders: &mut Vec<Order>) {
-    orders.retain(|order| {
-        let balance = order.order_meta_data.available_balance.unwrap_or_default();
-        !balance.is_zero()
-            && (order.order_creation.partially_fillable
-                || balance >= (order.order_creation.sell_amount + order.order_creation.fee_amount))
-    });
+// The order book has to make a choice for which orders to include when a user has multiple orders
+// selling the same token but not enough balance for all of them.
+// Assumes balance fetcher is already tracking all balances.
+fn solvable_orders(mut orders: Vec<Order>, balance_fetcher: &dyn BalanceFetching) -> Vec<Order> {
+    // Group orders by owner and sell token. For every group we get the balance that the orders
+    // share and pick orders prioritized by lower uid first to make it deterministic.
+    #[derive(Eq, PartialEq, Hash)]
+    struct Key {
+        owner: H160,
+        sell_token: H160,
+    }
+    let mut orders_map = HashMap::<Key, Vec<Order>>::new();
+    orders.sort_by_key(|order| order.order_meta_data.uid);
+    for order in orders {
+        let key = Key {
+            owner: order.order_meta_data.owner,
+            sell_token: order.order_creation.sell_token,
+        };
+        orders_map.entry(key).or_default().push(order);
+    }
+
+    let mut result = Vec::new();
+    for (key, orders) in orders_map {
+        let mut remaining_balance = match balance_fetcher.get_balance(key.owner, key.sell_token) {
```

Comment:
I think I should make another PR first so that we can return a different `Order` struct in the solvable_orders route than the orders route for the frontend. There we can decide how we want to encode balance for partial orders without affecting the old route.
Then I will come back to here and see what the best way to refactor the balance fetching is.

Reply:
I have refactored this code to get rid of the "warmed up" problem. Made an issue for the partially fillable orders because I think it's better to tackle that outside of this PR.